azurelinuxagent/common/protocol/extensions_goal_state.py

```diff
@@ -41,11 +52,56 @@ def __init__(self):
         self.required_features = []
 
     @staticmethod
-    def from_extensions_config(xml_text):
-        extensions_goal_state = ExtensionsGoalState()
-        extensions_goal_state.xml_text = xml_text
+    def create_empty():
+        return ExtensionsGoalState()
+
+    @staticmethod
+    def create_from_extensions_config(incarnation, xml_text):
+        return _ExtensionsGoalStateFromExtensionsConfig(incarnation, xml_text)
+
+    @staticmethod
+    def create_from_vm_settings(etag, json_text):
+        return _ExtensionsGoalStateFromVmSettings(etag, json_text)
+
+    def get_id(self):
+        """
+        Returns the incarnation number if the ExtensionsGoalState was created from ExtensionsConfig, or the etag if it
+        was created from vmSettings.
+        """
+        return self._id
+
+    def get_redacted_text(self):
+        """
+        Returns the raw text (either the ExtensionsConfig or the vmSettings) with any confidential data removed.
+        """
+        return None
+
+    @staticmethod
+    def compare(from_extensions_config, from_vm_settings):
+        """
+        Compares the current instance against 'other' and logs a GoalStateMismatch message if they are different
+        """
+        def report_difference(attribute):
+            message = "Mismatch in ExtensionsConfig (incarnation {0}) and vmSettings (etag {1}).\nAttribute: {2}\n{3} != {4}".format(
+                from_extensions_config.get_id(), from_vm_settings.get_id(),
+                attribute,
+                attrgetter(attribute)(from_extensions_config), attrgetter(attribute)(from_vm_settings)
+            )
+            logger.info("[GoalStateMismatch] {0}", message)
+            add_event(op=WALAEventOperation.GoalStateMismatch, is_success=False, message=message)
 
-        xml_doc = parse_doc(extensions_goal_state.xml_text)
+        if from_extensions_config.required_features != from_vm_settings.required_features:
+            report_difference("required_features")
+
+
+class _ExtensionsGoalStateFromExtensionsConfig(ExtensionsGoalState):
+    def __init__(self, incarnation, xml_text):
+        super(_ExtensionsGoalStateFromExtensionsConfig, self).__init__()
+        self._id = incarnation
+        self._text = xml_text
+        self._type = 'ExtensionsConfig'
```

Comment:
Maybe make the type an Enum - `ExtensionsConfig` vs `VmSettings`?

Reply:
done